File: LeetCode/BinaryTreeDFS/python/binaryTreeMaximumPathSum.py
from collections import deque
from typing import Dict, List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def __init__(self):
        self.res = 0

    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        #  Using a global variable and using an array for ease of manipulation
        self.res = [root.val]

        self.DFS(root)

        return self.res[0]
    
    def DFS(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        
        leftMax = self.DFS(root.left)
        rightMax = self.DFS(root.right)

        # DFS could return negative value, if negative, we want to use zero instead and not include the negative value
        leftMax = max(leftMax, 0)
        rightMax = max(rightMax, 0)

        # Compute the max path with split to both children
        self.res[0] = max(self.res[0], root.val + leftMax + rightMax)

        return root.val + max(leftMax, rightMax)

    @staticmethod
    def testSolution(input: Dict[str, List[str|int]]) -> None:
        # Create the Binary Tree from the input list using BFS
        binTree = Solution.BFSCreateBinaryTree(input["root"])

        # Print the Binary Tree and the expected output
        print("Input: " + Solution.BFSBinaryTreeToStr(binTree))
        print("Expected: " + str(input["expected"]))

        # Call the function to be tested
        res = Solution().maxPathSum(binTree)
   
        # Print the result
        print("Result: " + str(res))

        # Validate the result
        print("PASS" if res == input["expected"] else "FAIL")
        print("-" * 50)

    @staticmethod
    def BFSCreateBinaryTree(node: List[int|str]) -> TreeNode:
        arLen = len(node)

        if arLen < 1:
            return None

        root = TreeNode(node[0], None, None)
        i = 1

        # Using queue (BFS) to build the tree level by level
        queue = deque([root])

        while len(queue) > 0 and i < arLen:
            # Pre order: root, left, then right
            cur_node = queue.popleft()

            # Left child
            if i < arLen and node[i] != "null":
                cur_node.left = TreeNode(node[i], None, None)
                queue.append(cur_node.left)
            i += 1
            
            # Right child
            if i < arLen and node[i] != "null":
                cur_node.right = TreeNode(node[i], None, None)
                queue.append(cur_node.right)            
            i += 1

        return root
    
    @staticmethod
    def BFSBinaryTreeToStr(root: Optional[TreeNode]) -> None:

        if root is None:
            return "[]"
        
        sb = []

        # Using queue (BFS) to build the tree level by level
        queue = deque([root])


        while len(queue) > 0:
            curNode = queue.popleft()

            tmp = str(curNode.val) if curNode is not None else "null"

            sb.append(tmp)
 
            if curNode is not None:
                queue.append(curNode.left)
                queue.append(curNode.right)

        new_len = 0

        # Check starting from end of sb_array where the value is not "null"
        for i in range(len(sb) - 1, -1, -1):
            if sb[i] == "null":
                continue
            else:
                new_len = i
                break

        # Remove those extra "null"
        if len(sb) > 0:
            sb = sb[0 : new_len + 1]

        return "[" + ", ".join(sb) + "]"

    # @staticmethod
    # def validateResult(res: Optional[TreeNode], expected: Optional[TreeNode]) -> bool:
    #     resList = []
    #     expectedList = []

    #     qRes = deque([res])
    #     qExpected = deque([expected])

    #     while len(qRes) > 0:
    #         curNode = qRes.popleft()

    #         resList.append(curNode.val if curNode is not None else "null")

    #         if curNode is not None:
    #             qRes.append(curNode.left)
    #             qRes.append(curNode.right)

    #     while len(qExpected) > 0:
    #         curNode = qExpected.popleft()
            
    #         expectedList.append(curNode.val if curNode is not None else "null")

    #         if curNode is not None:
    #             qExpected.append(curNode.left)
    #             qExpected.append(curNode.right)

    #     # print(resList)
    #     # print(expectedList)

    #     return resList == expectedList

if __name__ == "__main__":
    input = {"root": [1,2,3], "expected": 6}
    Solution.testSolution(input)

    input = {"root": [-10,9,20,"null","null",15,7], "expected": 42}
    Solution.testSolution(input)

File: LeetCode/BinaryTreeDFS/java/binaryTreeMaximumPathSum/binaryTreeMaximumPathSum.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

/**
 * Definition for a binary tree node.
 **/
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

record binaryTreeMaximumPathSumRecord(String[] root, int expected) {
}

class binaryTreeMaximumPathSum {

    private int res;

    {
        this.res = Integer.MIN_VALUE;
    }

    public int maxPathSum(TreeNode root) {
        this.res = root.val;

        this.DFS(root);

        return this.res;
    }

    private int DFS(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int leftMax = this.DFS(root.left);
        int rightMax = this.DFS(root.right);

        leftMax = Math.max(leftMax, 0);
        rightMax = Math.max(rightMax, 0);

        this.res = Math.max(this.res, root.val + leftMax + rightMax);

        return root.val + Math.max(leftMax, rightMax);
    }

    public static void main(String[] args) {
        binaryTreeMaximumPathSumRecord input = new binaryTreeMaximumPathSumRecord(
                new String[] { "1", "2", "3" },
                6);
        binaryTreeMaximumPathSum.testSolution(input);

        input = new binaryTreeMaximumPathSumRecord(
                new String[] { "-10", "9", "20", "null", "null", "15", "7" },
                42);
        binaryTreeMaximumPathSum.testSolution(input);
    }

    private static void testSolution(binaryTreeMaximumPathSumRecord input) {
        // Create the binary tree from the input array using BFS
        TreeNode rootBTree = binaryTreeMaximumPathSum.BFSCreateBinaryTree(input.root());

        // Print the binary tree and the expected output
        System.out.println("Input: root: " + binaryTreeMaximumPathSum.BFSBinaryTreeToStr(rootBTree));
        System.out.println("Expected: " + input.expected());

        // Call the function to be tested
        int res = new binaryTreeMaximumPathSum().maxPathSum(rootBTree);

        // Print the result
        System.out.println("Result: " + res);

        // Validate the result
        System.out.println(res == input.expected() ? "PASS" : "FAIL");
        System.out.println("-".repeat(50));
    }

    private static TreeNode BFSCreateBinaryTree(String[] nodeValsArray) {
        int arLen = nodeValsArray.length;

        if (arLen < 1) {
            return null;
        }

        TreeNode root = new TreeNode(Integer.parseInt(nodeValsArray[0]), null, null);
        int i = 1;

        // Using queue (BFS) to build the tree level by level
        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.addLast(root);

        while (!queue.isEmpty() && i < arLen) {
            TreeNode curNode = queue.removeFirst();

            if (curNode != null) {
                // left child
                if (i < arLen && nodeValsArray[i] != "null") {
                    curNode.left = new TreeNode(Integer.parseInt(nodeValsArray[i]), null, null);
                    queue.addLast(curNode.left);
                }
                i++;

                // right child
                if (i < arLen && nodeValsArray[i] != "null") {
                    curNode.right = new TreeNode(Integer.parseInt(nodeValsArray[i]), null, null);
                    queue.addLast(curNode.right);
                }
                i++;
            }
        }

        return root;
    }

    private static String BFSBinaryTreeToStr(TreeNode root) {
        // Using BFS traversal to print tree value level by level
        if (root == null) {
            return "[]";
        }

        List<String> sb = new ArrayList<>();

        Deque<TreeNode> queue = new LinkedList<>(); // Use linkedList to allow addition of null object. ArrayDeque does
                                                    // not allow null object
        queue.addLast(root);

        while (!queue.isEmpty()) {
            TreeNode curNode = queue.removeFirst();
            sb.add(curNode != null ? String.valueOf(curNode.val) : "null");

            if (curNode != null) {
                queue.addLast(curNode.left);
                queue.addLast(curNode.right);
            }
        }

        while (!sb.isEmpty()) {
            if (sb.getLast() == "null") {
                sb.removeLast();
            } else {
                break;
            }
        }

        return sb.toString();
    }

    // private static boolean validateResult(TreeNode res, TreeNode expectedTree) {
    // List<String> resList = new ArrayList<>();
    // List<String> expectedList = new ArrayList<>();

    // // Use linkedList to allow addition of null object. ArrayDeque does
    // Deque<TreeNode> qRes = new LinkedList<>();
    // qRes.addLast(res);

    // Deque<TreeNode> qExpected = new LinkedList<>();
    // qExpected.addLast(expectedTree);

    // while (!qRes.isEmpty()) {
    // TreeNode curNode = qRes.pollFirst();

    // resList.add(curNode != null ? String.valueOf(curNode.val) : "null");

    // if (curNode != null) {
    // qRes.addLast(curNode.left);
    // qRes.addLast(curNode.right);
    // }
    // }

    // while (!qExpected.isEmpty()) {
    // TreeNode curNode = qExpected.pollFirst();

    // expectedList.add(curNode != null ? String.valueOf(curNode.val) : "null");

    // if (curNode != null) {
    // qExpected.addLast(curNode.left);
    // qExpected.addLast(curNode.right);
    // }
    // }

    // return expectedList.equals(resList);
    // }
}
